packages/datadog-plugin-mongodb-core/src/index.js

```diff
@@ -1,6 +1,121 @@
 'use strict'
 
-const unified = require('./unified')
-const legacy = require('./legacy')
+const Plugin = require('../../dd-trace/src/plugins/plugin')
+const { storage } = require('../../datadog-core')
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+class MongodbCorePlugin extends Plugin {
+  static get name () {
+    return 'mongodb-core'
+  }
 
-module.exports = [].concat(unified, legacy)
+  addSubs (func, start, asyncEnd = defaultAsyncEnd) {
+    this.addSub(`apm:mongodb:${func}:start`, start)
+    this.addSub(`apm:mongodb:${func}:end`, this.exit.bind(this))
+    this.addSub(`apm:mongodb:${func}:error`, errorHandler)
+    this.addSub(`apm:mongodb:${func}:async-end`, asyncEnd)
+  }
+
+  startSpan (customTags, store, ns, cmd, topology, operationName) {
+    const query = getQuery(cmd)
+    const resource = getResource(ns, query, operationName)
+    const tags = {
+      'service.name': this.config.service || `${this.tracer._service}-mongodb`,
+      'resource.name': resource,
+      'span.type': 'mongodb',
+      'span.kind': 'client',
+      'db.name': ns
+    }
+
+    for (const tag in customTags) {
+      tags[tag] = customTags[tag]
+    }
+    const span = this.tracer.startSpan('mongodb.query', {
+      childOf: store ? store.span : null,
+      tags
+    })
+
+    if (query) {
+      span.setTag('mongodb.query', query)
+    }
+
+    const options = topology && topology.s && topology.s.options
+
+    if (options && options.host && options.port) {
+      span.addTags({
+        'out.host': topology.s.options.host,
+        'out.port': topology.s.options.port
+      })
+    }
+
+    analyticsSampler.sample(span, this.config.measured)
+    return span
+  }
+
+  constructor (...args) {
+    super(...args)
+
+    this._addCommandSubs('command')
+    this._addCommandSubs('insert')
+    this._addCommandSubs('update')
+    this._addCommandSubs('remove')
+    this._addCommandSubs('getMore')
+    this._addCommandSubs('query')
+    this._addCommandSubs('_find')
+    this._addCommandSubs('killCursors')
+  }
+  _addCommandSubs (name) {
+    this.addSubs(name, ({ ns, ops, server, name }) => {
```

Comment:
Since the same callback is passed to `addSubs` on every invocation, it can just be part of `addSubs`, and you can replace this method entirely with addSubs.

Alternatively, and I think I prefer this, you can add all the subscriptions in a loop over an array of the commands.